app/services/users/remove_role.rb

```diff
@@ -0,0 +1,49 @@
+module Users
+  class RemoveRole
+    Response = Struct.new(:success, :error_message, keyword_init: true)
+
+    def self.call(*args)
+      new(*args).call
+    end
+
+    def initialize(user:, role:, resource_type:, admin:)
+      @user = user
+      @role = role
+      @resource_type = resource_type&.safe_constantize
+      @admin = admin
+      @response = Response.new(success: false)
+    end
+
+    def call
+      return response if super_admin_role?(role)
+      return response if user_is_current_user?(user)
+
+      if resource_type && user.remove_role(role, resource_type)
+        response.success = true
+      elsif user.remove_role(role)
+        response.success = true
+      else
+        response.error_message = "There was an issue removing this role. Please try again."
+      end
```

Comment:
Just a tiny detail, `if user.remove_role(role)` will be true even if the role doesn't exist (Rails returns the removed role in array if the role exists, it returns an empty array if the role does not exist), which might be what you want or not. Also keep in mind that `user.remove_role(:nonexistentrole)` will still work without raising an error, so I'm not sure there's a scenario for the `error_message`

A possible alternative to this conditional, could be:

```ruby
def call
  # ...
  if resource_type
    user.remove_role(role, resource_type)
  else
    user.remove_role(role)
  end
rescue StandardError => e
  response.error_message = ""There was an issue removing this role. #{e.message}"
  response.success = false
end
```

Reply:
Great point! I've updated this. 😊